core/builtin.py

```diff
@@ -235,6 +236,84 @@ def Echo(argv):
   return 0
 
 
+PRINTF_SPEC = _Register('printf')
+PRINTF_SPEC.ShortFlag('-v', args.Str)
+
+
+def Printf(argv, mem):
+  """
+  printf: printf [-v var] format [argument ...]
+  """
+  arg, args_consumed = PRINTF_SPEC.Parse(argv)
+  if args_consumed >= len(argv):
+    util.error('printf: need format string')
+    return 1
+  fmt = argv[args_consumed]
+  vals = argv[args_consumed + 1:]
+
+  result = ''
+  f = 0
+  v = 0
+  # Loop invariant: vals[v:] and fmt[f:] remain, to accumulate into result.
+  while True:
+    f_next = fmt.find('%', f)
+    if f_next < 0:
+      f_next = len(fmt)
+
+    result += fmt[f:f_next]  # TODO backslash-escapes, at least \n
+    f = f_next
+
+    if f >= len(fmt):
+      if v >= len(vals):
+        break
+      else:
+        # (handy!) bash printf quirk: re-use fmt to consume remaining vals.
+        f = 0
+        continue
+
+    c = fmt[f+1]
+    if c == '%':
+      f += 2
+      result += '%'
+      continue
+    elif c == 's':
+      f += 2
+      result += vals[v] if v < len(vals) else ''
+      v += 1
+      continue
+    elif c == 'q':
+      f += 2
+      result += _PrintfQuote(vals[v] if v < len(vals) else '')
+      v += 1
+      continue
+    elif c == 'd':
+      f += 2
+      val = vals[v] if v < len(vals) else '0'
+      v += 1
+      try:
+        num = int(val)
+      except ValueError:
+        # TODO should print message but carry on as if 0
+        util.error('printf: %s: invalid number' % (val,))
+        return 1
+      result += str(num)
+    else:
+      # TODO %b, %(fmt)T, plus "the standard ones in printf(1)"
+      raise NotImplementedError
+
+  if arg.v:
+    state.SetLocalString(mem, arg.v, result)
+  else:
+    sys.stdout.write(result)
+    sys.stdout.flush()
+  return 0
+
+
+def _PrintfQuote(val):
+  # TODO match specifics
+  return "'" + val.replace("'", "'\\''") + "'"
```

Comment:
I don't think this is right, because `\` isn't special within single quotes.

Reply:
Ah, it's a bit subtle, but I think it really is a correct implementation of "quote this for the shell" -- in fact it's the simplest one to write down explicitly that I know :-)

Consider a value like `FOO'BAR`. This function encodes that as `'FOO'\''BAR'` -- i.e. the single-quote gets encoded as a `\'` located *outside* any quoting, and everything else gets put into single quotes.

In other words, in that replacement string `"'\\''"` (i.e. the four characters `'\''`), the quote at the beginning *ends* a single-quoted string, and the one at the end starts a new one; they're there precisely to let the `\'` be written outside of any quotes.

Now, that said, this logic probably isn't the final version of this helper. It meets the spec in bash's `help printf`:
```
      %q	quote the argument in a way that can be reused as shell input
```
but it could e.g. skip quoting where none is needed. Bash's version does that, and also has a preference for backslashes:
```
bash$ printf %q$'\n' '"\$"'
\"\\\$\"

osh$ printf %q$'\n' '"\$"'
'"\$"'
```
I don't so much like the backslashes, but it's possible even those will turn out to be worth matching for compatibility.